torch/_library/triton.py

```diff
@@ -70,15 +93,118 @@ def visit_Call(self, node: ast.Call) -> None:
         collector = Visitor()
         collector.visit(tree)
         closure_vars = inspect.getclosurevars(fn)
-        resolved = []
-        # First, resolve triton kernel names
-        for name in collector.triton_kernels:
+
+        # build combined globals from closure and function's __globals__
+        all_globals: dict[str, Any] = {}
+        all_globals.update(closure_vars.globals)
+        if hasattr(fn, "__globals__"):
+            all_globals.update(fn.__globals__)
+
+        def extract_names_from_expr(expr: ast.expr) -> list[str]:
+            """Extract all Name references from an AST expression."""
+            names: list[str] = []
+
+            class NameExtractor(ast.NodeVisitor):
+                def visit_Name(self, node: ast.Name) -> None:
+                    names.append(node.id)
+
+                def visit_Call(self, node: ast.Call) -> None:
+                    # for function calls, visit the function and all args
+                    self.generic_visit(node)
+
+            NameExtractor().visit(expr)
+            return names
+
+        def resolve_to_kernel(obj: object) -> object | None:
+            """Check if obj is a triton kernel or wrapper and return the kernel."""
+            if isinstance(obj, (JITFunction, Autotuner)):
+                return obj
+            # handle wrappers that have a .fn attribute pointing to JITFunction
+            if callable(obj) and hasattr(obj, "fn"):
+                inner = obj.fn
+                if isinstance(inner, JITFunction):
+                    return inner
+            return None
+
+        def trace_to_global_kernels(
+            name: str, visited: set[str] | None = None
+        ) -> list[object]:
+            """
+            Trace a name through local assignments back to global triton kernels.
+
+            This handles patterns like:
+                kernel_fn = _my_kernel  # global
+                wrapped = wrapper(kernel_fn)
+                autotuned = autotune(wrapped)
+                capture_triton(autotuned)  # traces back to _my_kernel
+            """
+            if visited is None:
+                visited = set()
+
+            if name in visited:
+                return []
+
+            visited = visited | {name}
```

Comment:
Is it itentional that this clones away the visited set in the arg list? If so you may want to name the local variable different. If the goal was to mutate the visited set, it should be:
```suggestion
            visited.add(name)
```
or at least use
`|=`
Otherwise, it shouldn't overwrite the arg name.